### PR TITLE
Never crash when launching eos-tutorial

### PR DESCRIPTION
**Pull request by mcatanzaro**
https://phabricator.endlessm.com/T3727

### CODE REVIEW

File: js/ui/workspace.js [hunk 1]
```diff
@@ -1022,6 +1022,13 @@ const Workspace = new Lang.Class({
     positionWindows: function(flags) {
         this._positionWindowsFlags |= flags;
 
+        /* This might not be the best place for this check. It prevents the
+         * shell from crashing in _computeAllWindowSlots when launching the
+         * tutorial if the actor has not been added to the stage. See T3727.
+         */
+        if (!this.actor.mapped)
```
Comment (cosimoc) [inline js/ui/workspace.js]: Yeah, this does not seem the best place for the check...
What kind of backtrace did you get? Does the crash happen every time?

Comment (mcatanzaro) [inline js/ui/workspace.js]: Yes, the crash happens every time. Will reproduce it again and post the backtrace here tomorrow.

Comment (mcatanzaro) [inline js/ui/workspace.js]:
```
(eos-shell:2576): St-ERROR **: st_widget_get_theme_node called on the widget [0x8c6fbd8 StWidget.window-picker:last-child first-child] which is not in the stage.

Program received signal SIGTRAP, Trace/breakpoint trap.
g_logv (log_domain=0xb7dc0de8 "St", log_level=G_LOG_LEVEL_ERROR, format=0x0, 
    args=0x0)
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmessages.c:1086
1086	/usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmessages.c: No such file or directory.


(gdb) call gjs_dumpstack()
== Stack trace for context 0x8180c80 ==
Workspace<._computeAllWindowSlots@resource:///org/gnome/shell/ui/workspace.js:1645
wrapper@resource:///org/gnome/gjs/modules/lang.js:178
Workspace<._realPositionWindows@resource:///org/gnome/shell/ui/workspace.js:1055
wrapper@resource:///org/gnome/gjs/modules/lang.js:178
@resource:///org/gnome/shell/ui/workspace.js:1029


(gdb) bt
#0  0xffffffff in g_logv (log_domain=0xb7dc0de8 "St", log_level=G_LOG_LEVEL_ERROR, format=0x0, args=0x0)
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmessages.c:1086
#1  0xffffffff in g_log (log_domain=0xb7dc0de8 "St", log_level=G_LOG_LEVEL_ERROR, format=0xb7dc4fa0 "st_widget_get_theme_node called on the widget %s which is not in the stage.")
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmessages.c:1119
#2  0xffffffff in st_widget_get_theme_node (widget=0x8c6fbd8 [StWidget])
    at st/st-widget.c:651
#3  0xffffffff in ffi_call_SYSV () at ../src/x86/sysv.S:65
#4  0xffffffff in ffi_call (cif=0x869d0ec, fn=0xb7da8d10 <st_widget_get_theme_node>, rvalue=0xbfffc790, avalue=0xbfffc6b0) at ../src/x86/ffi.c:411
#5  0xffffffff in gjs_invoke_c_function(JSContext*, Function*, JSObject*, unsigned int, jsval*, jsval*, GArgument*) (context=context@entry=0x834e260, function=function@entry=0x869d0e0, obj=0xa4fc0e80, js_argc=0, js_argv=0x83acec0, js_rval=0xbfffc8d0, r_value=0x0) at gi/function.cpp:999
#6  0xffffffff in function_call(JSContext*, unsigned int, jsval*) (context=0x834e260, js_argc=0, vp=0x83aceb0) at gi/function.cpp:1322
#7  0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (args=..., native=<optimized out>, cx=0x834e260) at ./jscntxtinlines.h:321
#8  0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=c---Type <return> to continue, or q <return> to quit---
x@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:474
#9  0xffffffff in Interpret(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:2298
#10 0xffffffff in js::RunScript(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:438
#11 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=cx@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:500
#12 0xffffffff in js_fun_apply(JSContext*, unsigned int, JS::Value*) (cx=0x834e260, argc=2, vp=0xbfffd568)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jsfun.cpp:982
#13 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (args=..., native=<optimized out>, cx=0x834e260) at ./jscntxtinlines.h:321
#14 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=cx@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:481
#15 0xffffffff in js::Invoke(JSContext*, JS::Value const&, JS::Value const&, unsigned int, JS::Value*, JS::Value*) (cx=cx@entry=0x834e260, thisv=..., fval=..., argc=argc@entry=2, argv=argv@entry=0xbfffd744, rval=rval@entry=0xbfffd70c)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:531
---Type <return> to continue, or q <return> to quit---
#16 0xffffffff in js::jit::DoCallFallback(JSContext*, js::jit::BaselineFrame*, js::jit::ICCall_Fallback*, uint32_t, JS::Value*, JS::MutableHandleValue) (cx=0x834e260, frame=0xbfffd79c, stub=0x91ef840, argc=2, vp=0xbfffd734, res=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineIC.cpp:7007
#17 0xffffffff in  ()
#18 0x091ef840 in  ()
#19 0xffffffff in  ()
#20 0xffffffff in EnterBaseline(JSContext*, js::jit::EnterJitData&) (cx=0x0, 
    cx@entry=0x834e260, data=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineJIT.cpp:105
#21 0xffffffff in js::jit::EnterBaselineMethod(JSContext*, js::RunState&) (data=..., cx=0x834e260)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineJIT.cpp:81
#22 0xffffffff in js::jit::EnterBaselineMethod(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineJIT.cpp:136
#23 0xffffffff in Interpret(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:2334
#24 0xffffffff in js::RunScript(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:438
#25 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=c---Type <return> to continue, or q <return> to quit---
x@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:500
#26 0xffffffff in js_fun_apply(JSContext*, unsigned int, JS::Value*) (cx=0x834e260, argc=2, vp=0xbfffe3a8)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jsfun.cpp:982
#27 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (args=..., native=<optimized out>, cx=0x834e260) at ./jscntxtinlines.h:321
#28 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=cx@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:481
#29 0xffffffff in js::Invoke(JSContext*, JS::Value const&, JS::Value const&, unsigned int, JS::Value*, JS::Value*) (cx=cx@entry=0x834e260, thisv=..., fval=..., argc=argc@entry=2, argv=argv@entry=0xbfffe584, rval=rval@entry=0xbfffe54c)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:531
#30 0xffffffff in js::jit::DoCallFallback(JSContext*, js::jit::BaselineFrame*, js::jit::ICCall_Fallback*, uint32_t, JS::Value*, JS::MutableHandleValue) (cx=0x834e260, frame=0xbfffe5dc, stub=0x91ef840, argc=2, vp=0xbfffe574, res=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineIC.cpp:7007
#31 0xffffffff in  ()
#32 0x091ef840 in  ()
#33 0xffffffff in  ()
#34 0xffffffff in EnterBaseline(JSContext*, js::jit::EnterJitData&) (cx=0x0, 
    cx@entry=0x834e260, data=...)
---Type <return> to continue, or q <return> to quit---
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineJIT.cpp:105
#35 0xffffffff in js::jit::EnterBaselineMethod(JSContext*, js::RunState&) (data=..., cx=0x834e260)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineJIT.cpp:81
#36 0xffffffff in js::jit::EnterBaselineMethod(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jit/BaselineJIT.cpp:136
#37 0xffffffff in Interpret(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:2334
#38 0xffffffff in js::RunScript(JSContext*, js::RunState&) (cx=cx@entry=0x834e260, state=...)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:438
#39 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=cx@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:500
#40 0xffffffff in js::CallOrConstructBoundFunction(JSContext*, unsigned int, JS::Value*) (cx=0x834e260, argc=0, vp=0xbffff0e8)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jsfun.cpp:1212
#41 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (args=..., native=<optimized out>, cx=0x834e260) at ./jscntxtinlines.h:321
#42 0xffffffff in js::Invoke(JSContext*, JS::CallArgs, js::MaybeConstruct) (cx=cx@entry=0x834e260, args=..., construct=construct@entry=js::NO_CONSTRUCT)
---Type <return> to continue, or q <return> to quit---
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:481
#43 0xffffffff in js::Invoke(JSContext*, JS::Value const&, JS::Value const&, unsigned int, JS::Value*, JS::Value*) (cx=cx@entry=0x834e260, thisv=..., fval=..., argc=argc@entry=0, argv=argv@entry=0xbffff1c0, rval=rval@entry=0xbffff228)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/vm/Interpreter.cpp:531
#44 0xffffffff in JS_CallFunctionValue(JSContext*, JSObject*, JS::Value, unsigned int, JS::Value*, JS::Value*) (cx=0x834e260, objArg=0x0, fval=..., argc=0, argv=0xbffff1c0, rval=0xbffff228)
    at /usr/src/packages/BUILD/mozjs24-24.2.0/js/src/jsapi.cpp:5639
#45 0xffffffff in gjs_callback_closure(ffi_cif*, void*, void**, void*) (cif=0x925bb58, result=0xbffff390, args=0xbffff330, data=0x925bb40)
    at gi/function.cpp:284
#46 0xffffffff in ffi_closure_SYSV_inner (closure=0xa72288a0, respp=0xbffff39c, args=0xbffff3b0) at ../src/x86/ffi.c:496
#47 0xffffffff in ffi_closure_SYSV () at ../src/x86/sysv.S:199
#48 0xffffffff in run_repaint_laters (data=0x0) at core/util.c:821
#49 0xffffffff in _clutter_run_repaint_functions (flags=CLUTTER_REPAINT_FLAGS_PRE_PAINT) at clutter-main.c:3427
#50 0xffffffff in clutter_clock_dispatch (master_clock=<optimized out>, stages=<optimized out>) at clutter-master-clock-default.c:437
#51 0xffffffff in clutter_clock_dispatch (source=0x8d0c630, callback=0x0, user_data=0x0) at clutter-master-clock-default.c:567
#52 0xffffffff in g_main_context_dispatch (context=<optimized out>)
---Type <return> to continue, or q <return> to quit---
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmain.c:3154
#53 0xffffffff in g_main_context_dispatch (context=0x0)
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmain.c:3769
#54 0xffffffff in g_main_context_iterate (context=0x8070e50, block=0, 
    block@entry=1, dispatch=1, self=<optimized out>)
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmain.c:3840
#55 0xffffffff in g_main_loop_run (loop=0x805bae0)
    at /usr/src/packages/BUILD/glib2.0-2.46.2+dev28.d2e1d0a/./glib/gmain.c:4034
#56 0xffffffff in meta_run () at core/main.c:555
#57 0x08049443 in main (argc=1, argv=0xbffff6b4) at main.c:399
```

**Comment by cosimoc** [comment]
@mcatanzaro I checked the upstream code and it has changed quite significantly from what we have.
However, they did add some `actor.mapped` checks; does adding it here [1] fix the issue? If so, please update the patch. In other words, it would be nice to track down where the call to `positionWindows()` comes from.

[1] https://github.com/endlessm/eos-desktop/blob/master/js/ui/workspace.js#L974

**Comment by mcatanzaro** [comment]
No, adding the check in setGeometry does not fix the issue. I actually tried that first, after looking at the upstream changes, but it didn't help. Unfortunately there are so many upstream changes that it's not clear 

I will investigate further to find out where the call to positionWindows() is coming from. It's a shame gjs_dumpstack does not show more.

Anyway, I am going to close this pull request and reopen it on a T3881 branch, since I found that bug report matches this problem exactly.

**Comment by mcatanzaro** [comment]
It's coming from the very bottom of _doAddWindow().

**Comment by mcatanzaro** [comment]
The call chain here is:

_windowEnteredMonitor -> _doAddWindow -> calls _doAddWindow again in an idle -> positionWindows -> realPositionWindows -> computeAllWindowSlots -> crash

To fix this, we have to ensure the Workspace object's actor is mapped before calling computeAllWindows. An early return in either positionWindows or realPositionWindows seems to work well; the code would always crash in that case currently.

I'm not sure there's a better place for this check. Upstream has added checks in a bunch of earlier places, but the code has diverged so significantly I think it's safer to stick with the check in positionWindows. What do you think, Cosimo?

**Comment by cosimoc** [comment]
Sounds good, let's merge this as-is for now.

**Pull request merged**